core/src/test/java/org/elasticsearch/search/aggregations/UnmappedFieldValidationIT.java

```diff
@@ -0,0 +1,121 @@
+        assertAcked(prepareCreate("idx_unmapped").addMapping("type", "bar", "type=keyword").get());
+        indexRandom(true,
+                client().prepareIndex("idx", "type").setSource("foo", "1"),
+                client().prepareIndex("idx", "type").setSource("foo", "2"),
+                client().prepareIndex("idx", "type").setSource("foo", "3"),
+                client().prepareIndex("idx", "type").setSource("foo", "4"),
+                client().prepareIndex("idx", "type").setSource("foo", "5"),
+                client().prepareIndex("idx_unmapped", "type").setSource("bar", "1"),
+                client().prepareIndex("idx_unmapped", "type").setSource("bar", "2"),
+                client().prepareIndex("idx_unmapped", "type").setSource("bar", "3"),
+                client().prepareIndex("idx_unmapped", "type").setSource("bar", "4"),
+                client().prepareIndex("idx_unmapped", "type").setSource("bar", "5")
+                );
+    }
+
+    public void testUnmappedTerms() {
+        
+        SearchResponse response = client().prepareSearch("idx", "idx_unmapped")
+                .addAggregation(terms("my_terms").field(randomFrom("foo", "bar"))).get();
+        assertSearchResponse(response);
+        Terms terms = response.getAggregations().get("my_terms");
+        assertEquals(5, terms.getBuckets().size());
+    }
+    
+    public void testStrictUnmappedStringTermsSingleIndex() {
+        try {
+            client().prepareSearch("idx_unmapped").setBatchedReduceSize(2).setStrictFieldNames(true)
+                .addAggregation(terms("my_terms1").field("foo")).get();
+            fail("All unmapped fields failure expected");
+        } catch (SearchPhaseExecutionException e) {
+            assertTrue(e.getCause().getMessage().contains("The following fields were unmapped in all indices searched: [foo]"));
+        }
+    }
+    
+    public void testStrictUnmappedStringTermsMultiIndex() {
+        try {
+            client().prepareSearch("idx","idx_unmapped").setStrictFieldNames(true)
+                .addAggregation(terms("my_terms").field("all_unmapped_str")).get();
+            fail("All unmapped fields failure expected");
+        } catch (SearchPhaseExecutionException e) {
+            assertTrue(e.getCause().getMessage().contains(
+                    "The following fields were unmapped in all indices searched: [all_unmapped_str]"));
+        }
+    }  
+    
+    public void testStrictUnmappedStringTermsMultiIndexMultiReduce() {
+        try {
+            client().prepareSearch("idx","idx_unmapped").setBatchedReduceSize(2).setStrictFieldNames(true)
+                .addAggregation(terms("my_terms").field("all_unmapped_str")).get();
+            fail("All unmapped fields failure expected");
+        } catch (SearchPhaseExecutionException e) {
+            assertTrue(e.getCause().getMessage().contains(
+                    "The following fields were unmapped in all indices searched: [all_unmapped_str]"));
+        }
```

Comment:
FYI we usually prefer expectThrows which avoids some of the traps of the try/fail/catch/assert idiom